### PR TITLE
feat: update x/bankplus for deprecation

### PR DESCRIPTION
**Pull request by jaeseung-bae**
## Description

closes: #1212
* **NOTE: This PR has proto breaking change. `lbm/bankplus/v1/bankplus.proto` has been deleted for deprecation.**

## Checklist:

- [x] I followed the [contributing guidelines](https://github.com/Finschia/finschia-sdk/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/Finschia/finschia-sdk/blob/main/CODE_OF_CONDUCT.md).
- [x] I have added a relevant changelog to `CHANGELOG.md`
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated API documentation `client/docs/swagger-ui/swagger.yaml`

### CODE REVIEW

File: x/bankplus/keeper/inactive.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+func (keeper BaseKeeper) isStoredInactiveAddr(ctx context.Context, address sdk.AccAddress) bool {
+	store := keeper.storeService.OpenKVStore(ctx)
+	bz, _ := store.Get(inactiveAddrKey(address))
+	return bz != nil
+}
+
+// addToInactiveAddr adds a blocked address to the store.
+func (keeper BaseKeeper) addToInactiveAddr(ctx context.Context, address sdk.AccAddress) {
+	store := keeper.storeService.OpenKVStore(ctx)
+	blockedCAddr := types.InactiveAddr{Address: address.String()}
```
Comment (0Tech) [inline x/bankplus/keeper/inactive.go:32]: We should not use the global codec.

**Review by 0Tech — CHANGES_REQUESTED** [review]

File: x/bankplus/keeper/keeper_test.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	s.mockAuthKeeperFor(mintAcc)
+	s.mockAuthKeeperFor(burnerAcc)
+
+	s.bkPlusKeeper = keeper.NewBaseKeeper(
+		encCfg.Codec,
+		storeService,
+		s.authKeeper,
+		map[string]bool{},
+		true,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:101]: Please use keeper codec

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -323,3 +323,34 @@ jobs:
...
+          cd x/bankplus
+          go test -mod=readonly -timeout 30m -coverprofile=coverage.out -covermode=atomic -tags='norace ledger test_ledger_mock rocksdb_build' ./...
+      - name: sonarcloud
+        if: ${{ env.GIT_DIFF && !github.event.pull_request.draft && env.SONAR_TOKEN != null }}
+        uses: SonarSource/sonarcloud-github-action@master
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        with:
+          projectBaseDir: x/bankplus/
```
Comment (ulbqb) [inline .github/workflows/test.yml:356]:
```suggestion
          projectBaseDir: x/bankplus/

```

File: x/bankplus/keeper/keeper_test.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+func (s *IntegrationTestSuite) TestInitializeBankPlus() {
+	newKeeper := keeper.NewBaseKeeper(
+		s.codec,
+		s.storeService,
+		s.authKeeper,
+		map[string]bool{},
+		true,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:207]: ditto

File: x/bankplus/keeper/keeper_test.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+	newKeeper.AddToInactiveAddr(s.ctx, blockedAcc.GetAddress())
+	s.Require().True(newKeeper.IsInactiveAddr(blockedAcc.GetAddress()))
+
+	anotherNewKeeper := keeper.NewBaseKeeper(
+		s.codec,
+		s.storeService,
+		s.authKeeper,
+		map[string]bool{},
+		true,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:219]: ditto

File: x/bankplus/keeper/keeper_test.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	s.Require().True(anotherNewKeeper.IsInactiveAddr(blockedAcc.GetAddress()))
+}
+
+func (s *IntegrationTestSuite) TestSendCoinsFromModuleToAccount_Blacklist() {
+	addr1 := sdk.AccAddress([]byte("addr1_______________"))
+	newKeeper := keeper.NewBaseKeeper(
+		s.codec,
+		s.storeService,
+		s.authKeeper,
+		map[string]bool{addr1.String(): true},
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:232]: ditto

File: x/bankplus/keeper/keeper_test.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+		true,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:234]: ditto

File: x/bankplus/keeper/keeper_test.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+	outputs := []types.Output{types.NewOutput(burnerAcc.GetAddress(), initCoins)}
+
+	targetKeeper := func(isDeact bool) keeper.BaseKeeper {
+		return keeper.NewBaseKeeper(
+			s.codec,
+			s.storeService,
+			s.authKeeper,
+			make(map[string]bool),
+			isDeact,
+			authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
Comment (ulbqb) [inline x/bankplus/keeper/keeper_test.go:257]: ditto

**Comment by ulbqb** [comment]
Some ci tests are failing.

**Comment by jaeseung-bae** [comment]
> Some ci tests are failing.
Yes, I'm working on it but I have not found any clue, especially `test-e2e` failure.
After providing bankplus moudle for DI instead of original bank, tests in `tests/e2e` begin to fail.
Any help would be appreciated.

File: x/bankplus/keeper/keeper.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+// SendCoins transfers amt coins from a sending account to a receiving account.
+// This is wrapped bank the `SendKeeper` interface of `bank` module,
+// and checks if `toAddr` is a inactiveAddr managed by the module.
+func (keeper BaseKeeper) SendCoins(ctx context.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) error {
+	// if toAddr is smart contract, check the status of contract.
+	if keeper.isInactiveAddr(toAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", toAddr)
+	}
+
+	return keeper.BaseSendKeeper.SendCoins(ctx, fromAddr, toAddr, amt)
```
Comment (0Tech) [inline x/bankplus/keeper/keeper.go:126]: We can remove duplicate logic using the send restriction feature. This, or another issue should address it.

File: x/bankplus/keeper/keeper.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	senderAddr := keeper.ak.GetModuleAddress(senderModule)
+	if senderAddr.Empty() {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", senderModule))
+	}
+
+	if keeper.BlockedAddr(recipientAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", recipientAddr)
+	}
+
+	return keeper.SendCoins(ctx, senderAddr, recipientAddr, amt)
```
Comment (0Tech) [inline x/bankplus/keeper/keeper.go:75]: It would be better move the modification of `SendCoins()` to this method. It may also apply to other methods.

**Review by ulbqb — DISMISSED** [review]
LGTM

File: x/bankplus/keeper/keeper.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+// and checks if `toAddr` is a inactiveAddr managed by the module.
+func (k BaseKeeper) SendCoins(ctx context.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) error {
+	// if toAddr is smart contract, check the status of contract.
+	if k.isInactiveAddr(toAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", toAddr)
+	}
+
+	return k.BaseSendKeeper.SendCoins(ctx, fromAddr, toAddr, amt)
+}
+
```
Comment (0Tech) [inline x/bankplus/keeper/keeper.go:89]: Could you also remove this method?

Comment (jaeseung-bae) [reply x/bankplus/keeper/keeper.go:89]: I'll check again

**Comment by jaeseung-bae** [comment]
Please don't review this PR.
I'm thinking of removing bankplus while keeping it's role from other place.

**Comment by jaeseung-bae** [comment]
Please skip failure of `test-sim-nondeterminism`. Another PR will fix this.

File: x/bankplus/keeper/keeper.go (hunk 11 above)
Comment (jaeseung-bae) [reply x/bankplus/keeper/keeper.go]: Removed

**Comment by ulbqb** [comment]
Where SendRestrictionFn is set in the future?

**Comment by jaeseung-bae** [comment]
> Where SendRestrictionFn is set in the future? And can't SendRestrictionFn use inactive address data in bank kvstore as it is?

WASMD will set `SendRestrictionFn`. You're right, inactive address data will be removed and can't use it . Fortunately WASMD has the same inactive address in it's store.

File: simapp/app_config.go [hunk 12]
```diff
@@ -48,7 +48,6 @@ import (
 	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
 	"github.com/cosmos/cosmos-sdk/x/authz"
 	_ "github.com/cosmos/cosmos-sdk/x/authz/module" // import for side-effects
-	_ "github.com/cosmos/cosmos-sdk/x/bank"         // import for side-effects
```
Comment (ulbqb) [inline simapp/app_config.go:51]: Can't this be used? I think it's enough to remove inactive address in bankplus module and return empty output on ProvideModule.

Comment (jaeseung-bae) [reply simapp/app_config.go:51]: `_ "github.com/cosmos/cosmos-sdk/x/bank"` can be added. 
Bankplus already imports the original bank package, so it has no effect to add the import or not.

Is it okay to return empty output in bankplus's provider??
I'm not sure, but it needs to provide bankplus'  appmodule to register migration logic(removing inactive addresses).

**Review by 0Tech — COMMENTED** [review]
We should consider implementing the migration on the app. You may try to implement it on simapp, and if it works, you can open the corresponding PR on Finschia/finschia.

File: simapp/app_config.go (hunk 12 above)
Comment (ulbqb) [reply simapp/app_config.go:51]: Of cause, bankplus's appmodule is needed but I think it doesn't need to provide bank keeper.

**Comment by ulbqb** [comment]
How about leaving deprecator in bankplus because it's used by finschia/finschia? And proto files shouldn't be removed.

**Comment by jaeseung-bae** [comment]
> How about leaving deprecator in bankplus because it's used by finschia/finschia? And proto files shouldn't be removed.

I'm thinking of closing this PR when another PR on finschia/finschia is ready.
I'll make another PR just for bankplus deprecation logic on finschia/finschia.
finschia-sdk will look like it's never had bankplus before.

**Comment by ulbqb** [comment]
In my opinion, the repo that has better to have deprecator is finschia-sdk not finschia. If finschia, that makes it hard to follow history of bankplus and that makes proto files scattered.

**Comment by jaeseung-bae** [comment]
> tter to have deprecator is finschia-sdk not finschia. If finschia, that makes it hard to follow history of bankplus and that makes proto files scattered.

I agree. It seems better to keep bankplus logic here for history.

**Review by ulbqb — CHANGES_REQUESTED** [review]
Is CI testing bankplus? I suppose setting .github/workflow/test.yml is needed.

File: x/bankplus/deprecator.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+//		for _, key := range app.kvStoreKeys() {
+//			if key.Name() == banktypes.StoreKey {
+//				err := internal.DeprecateBankPlus(ctx, key)
+//				if err != nil {
+//					panic(fmt.Errorf("failed to deprecate x/bankplus: %w", err))
+//				}
+//			}
+//		}
+//	}
+func DeprecateBankPlus(ctx context.Context, bankKey *storetypes.KVStoreKey) error {
```
Comment (ulbqb) [inline x/bankplus/deprecator.go:46]: https://github.com/cosmos/cosmos-sdk/blob/v0.50.2/x/bank/migrations/v4/store.go#L20C49-L20C69

How about using `store.KVStoreService` instead of `storetypes.KVStoreKey` like migration?

**Review by ulbqb — APPROVED** [review]
LGTM

File: x/bankplus/deprecator.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+	adapter := runtime.KVStoreAdapter(kvStore)
+	iter := storetypes.KVStorePrefixIterator(adapter, inactiveAddrsKeyPrefix)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		err := kvStore.Delete(iter.Key())
+		if err != nil {
+			return err
+		}
+	}
```
Comment (0Tech) [inline x/bankplus/deprecator.go:59]: Is it OK to delete keys while iterating the very range?

Comment (jaeseung-bae) [reply x/bankplus/deprecator.go:59]: let me check.

Comment (jaeseung-bae) [reply x/bankplus/deprecator.go:59]: I think It's better to modify remove logic regardless of current or future behavior of iterator.

**Review by ulbqb — APPROVED** [review]
